[SPARK-14720][SPARK-13643] Remove HiveContext (step 1)

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

In Spark 2.0 we will have a new entry point for users known as the `SparkSession`. This class will handle the lazy initialization of the Hive metastore if the user runs commands that require interaction with the metastore (e.g. `CREATE TABLE`). With this, we can remove the `HiveContext`, which is an odd API to be exposed to Spark users.

This patch doesn't fully remove `HiveContext` but does most of the work. A follow-up patch will actually delete the file itself. I've left all the variable naming and any further refactor for later.
## How was this patch tested?

Jenkins.

Commit:
[SPARK-14647][SQL] Group SQLContext/HiveContext state into SharedState

Commit:
Make HiveSessionState take in SQLContext, not HiveContext

This requires changing all the downstream places that take in
HiveContext and replacing that with (SQLContext, HiveSessionState).

Commit:
Move QueryExecution out of HiveContext

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

Commit:
Minor cleanup

Commit:
Implement SparkSession and use it to track state

Now both shared state and session state is tracked in SparkSession
and we use reflection to instantiate them. After this commit
SQLContext and HiveContext are just wrappers for SparkSession.

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

Commit:
Clean up some TODO's and bad signatures

Commit:
Move the bulk of HiveContext into SessionCatalog

Commit:
Remove more things from HiveContext

Commit:
Fix style

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

SparkQA commented:
**[Test build #56166 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56166/consoleFull)** for PR 12485 at commit [`ce1214d`](https://github.com/apache/spark/commit/ce1214d79a90ff212ad871929ed1031350dd8f68).

Commit:
Minor fixes

Commit:
Use in-memory catalog by default in tests

SparkQA commented:
**[Test build #56170 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56170/consoleFull)** for PR 12485 at commit [`6019541`](https://github.com/apache/spark/commit/60195413d268e5ff7f1c295cc6cbc60794406619).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -132,7 +134,7 @@ class SQLContext private[sql](
+  def setConf(props: Properties): Unit = sessionState.setConf(props)
 
   /** Set the given Spark SQL configuration property. */
   private[sql] def setConf[T](entry: ConfigEntry[T], value: T): Unit = conf.setConf(entry, value)
```
Seems we also need to change this?

Commit:
Fix NPE when initializing HiveSessionState

SparkQA commented:
**[Test build #56174 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56174/consoleFull)** for PR 12485 at commit [`75d1115`](https://github.com/apache/spark/commit/75d1115e140b8401343c0b96429ccd346e3a4b17).

Commit:
Fix the conf

SparkQA commented:
**[Test build #56166 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56166/consoleFull)** for PR 12485 at commit [`ce1214d`](https://github.com/apache/spark/commit/ce1214d79a90ff212ad871929ed1031350dd8f68).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56166/
Test FAILed.

SparkQA commented:
**[Test build #56170 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56170/consoleFull)** for PR 12485 at commit [`6019541`](https://github.com/apache/spark/commit/60195413d268e5ff7f1c295cc6cbc60794406619).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56170/
Test FAILed.

SparkQA commented:
**[Test build #56174 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56174/consoleFull)** for PR 12485 at commit [`75d1115`](https://github.com/apache/spark/commit/75d1115e140b8401343c0b96429ccd346e3a4b17).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56174/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

Commit:
Fix REPL in in-memory case

Previously we still tried to load HiveContext even if the user
explicitly specified an "in-memory" catalog impelmentation. Now
it will load a SQLContext in this case.

Commit:
Fix tests: set "in-memory" in more places

Commit:
Fix style

SparkQA commented:
**[Test build #56249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56249/consoleFull)** for PR 12485 at commit [`0d7309b`](https://github.com/apache/spark/commit/0d7309b20f72809659306aec47f29c8585f36dd7).

SparkQA commented:
**[Test build #56251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56251/consoleFull)** for PR 12485 at commit [`575fd47`](https://github.com/apache/spark/commit/575fd478998cd9c970342c85402582a6dd82cd51).

Commit:
Fix SQLExecutionSuite

It was failing because we were passing in a subclass of
SparkContext into SparkSession, and the reflection was using
the wrong class to get the constructor. This is now fixed with
ClassTags.

SparkQA commented:
**[Test build #56254 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56254/consoleFull)** for PR 12485 at commit [`bc35206`](https://github.com/apache/spark/commit/bc352068a40a13712e0f16bbc81d66486e4a3635).

Commit:
Fix ParquetHadoopFsRelationSuite

Avoid some unnecessary casts.

SparkQA commented:
**[Test build #56251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56251/consoleFull)** for PR 12485 at commit [`575fd47`](https://github.com/apache/spark/commit/575fd478998cd9c970342c85402582a6dd82cd51).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56251/
Test FAILed.

SparkQA commented:
**[Test build #56249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56249/consoleFull)** for PR 12485 at commit [`0d7309b`](https://github.com/apache/spark/commit/0d7309b20f72809659306aec47f29c8585f36dd7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56249/
Test FAILed.

SparkQA commented:
**[Test build #56254 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56254/consoleFull)** for PR 12485 at commit [`bc35206`](https://github.com/apache/spark/commit/bc352068a40a13712e0f16bbc81d66486e4a3635).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56254/
Test FAILed.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -0,0 +1,100 @@
+   */
+  private def reflect[T, Arg <: AnyRef](
+      className: String,
+      ctorArg: Arg)(implicit ctorArgTag: ClassTag[Arg]): T = {
```
You don't need a class tag.  You can call `ctorArg.getClass()`.

andrewor14 replied:
didn't work because there are places where we pass subclasses of `Arg` in here

marmbrus commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -89,4 +89,11 @@ package object config {
+  private[spark] val CATALOG_IMPLEMENTATION = ConfigBuilder("spark.sql.catalogImplementation")
+    .internal()
+    .stringConf
+    .checkValues(Set("hive", "in-memory"))
```
Should this be a class name?

andrewor14 replied:
we use this for both `SharedState` and `SessionState`. If we want to use a class name then we need two configs.

Commit:
Fix HiveUDFSuite + refactor TestHive

The problem was that we weren't using the right QueryExecution
when we called TestHive.sessionState.executePlan. We were using
HiveQueryExecution instead of the custom one that we created
in TestHiveContext.

This turned out to be very difficult to fix due to the tight
coupling of QueryExecution within TestHiveContext. I had to
refactor this code significantly to extract the nested logic
one by one.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56271/
Test FAILed.

Commit:
Make diff slightly smaller?

SparkQA commented:
**[Test build #56273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56273/consoleFull)** for PR 12485 at commit [`303f991`](https://github.com/apache/spark/commit/303f991576f0a082b2d8cb4d98e9a866bb7d07a1).

Commit:
Fix test compile

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

SparkQA commented:
**[Test build #56276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56276/consoleFull)** for PR 12485 at commit [`b3d23fa`](https://github.com/apache/spark/commit/b3d23fa8892ead2508f449063fe8874ab3ebb47d).

SparkQA commented:
**[Test build #56276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56276/consoleFull)** for PR 12485 at commit [`b3d23fa`](https://github.com/apache/spark/commit/b3d23fa8892ead2508f449063fe8874ab3ebb47d).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait PartitionCoalescer `
  * `class PartitionGroup(val prefLoc: Option[String] = None) `
  * `trait ObjectProducer extends LogicalPlan `
  * `trait ObjectConsumer extends UnaryNode `
  * `case class DeserializeToObject(`
  * `case class SerializeFromObject(`
  * `case class AppendColumnsWithObject(`
  * `case class AppendColumnsWithObject(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56276/
Test FAILed.

Commit:
Fix HiveResolutionSuite

SparkQA commented:
**[Test build #56278 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56278/consoleFull)** for PR 12485 at commit [`ddc752a`](https://github.com/apache/spark/commit/ddc752a134cdea5b06386b0d4a3e7436dc5ba777).

Commit:
Fix StatisticsSuite

Commit:
Minor change

SparkQA commented:
**[Test build #56273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56273/consoleFull)** for PR 12485 at commit [`303f991`](https://github.com/apache/spark/commit/303f991576f0a082b2d8cb4d98e9a866bb7d07a1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56273/
Test FAILed.

SparkQA commented:
**[Test build #56278 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56278/consoleFull)** for PR 12485 at commit [`ddc752a`](https://github.com/apache/spark/commit/ddc752a134cdea5b06386b0d4a3e7436dc5ba777).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56278/
Test FAILed.

Commit:
Fix HiveUDFSuite (and many others)

The problem was that we were getting everything from
executionHive's hiveconf and setting that in metadataHive,
overriding the value of `hive.metastore.warehouse.dir`,
which we customize in TestHive. This resulted in a bunch
of "Table src does not exist" errors from Hive.

SparkQA commented:
**[Test build #56303 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56303/consoleFull)** for PR 12485 at commit [`8bf1236`](https://github.com/apache/spark/commit/8bf123601db1d33455f82d30aa19bdd875bd00d7).

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

SparkQA commented:
**[Test build #56305 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56305/consoleFull)** for PR 12485 at commit [`74b105e`](https://github.com/apache/spark/commit/74b105e59a9ec8e529d711f09199fda86ee8f7d3).

SparkQA commented:
**[Test build #56303 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56303/consoleFull)** for PR 12485 at commit [`8bf1236`](https://github.com/apache/spark/commit/8bf123601db1d33455f82d30aa19bdd875bd00d7).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56303/
Test FAILed.

SparkQA commented:
**[Test build #56305 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56305/consoleFull)** for PR 12485 at commit [`74b105e`](https://github.com/apache/spark/commit/74b105e59a9ec8e529d711f09199fda86ee8f7d3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Timer(val iteration: Int) `
  * `  case class Case(name: String, fn: Timer => Unit)`
  * `trait ScalaReflection `
  * `trait CheckAnalysis extends PredicateHelper `
  * `abstract class SubqueryExpression extends Expression `
  * `abstract class PredicateSubquery extends SubqueryExpression with Unevaluable with Predicate `
  * `case class InSubQuery(value: Expression, query: LogicalPlan) extends PredicateSubquery `
  * `case class Exists(query: LogicalPlan) extends PredicateSubquery `
  * `case class FilePartition(index: Int, files: Seq[PartitionedFile]) extends RDDPartition`
  * `abstract class OutputWriterFactory extends Serializable `
  * `abstract class OutputWriter `
  * `case class HadoopFsRelation(`
  * `trait FileFormat `
  * `case class Partition(values: InternalRow, files: Seq[FileStatus])`
  * `trait FileCatalog `
  * `class HDFSFileCatalog(`
  * `  case class FakeFileStatus(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56305/
Test FAILed.

rxin commented:
Hey @andrewor14 - I took a quick look at this. Shouldn't we move the sessionstate stuff in its own pr since it would be easier to get that in and also easier to review?

yhuai committed:
Fix tests

Commit:
Merge branch 'master' of github.com:apache/spark into spark-session

SparkQA commented:
**[Test build #56374 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56374/consoleFull)** for PR 12485 at commit [`9422128`](https://github.com/apache/spark/commit/9422128c6bd9c3ffbdf728eb6fd82eb17c50aef6).

yhuai committed:
Fix Python tests.

It may take time to track all places where we only use SQLContext.
So, let's change the catalog conf's default value to in-memory.
In the constructor of HiveContext, we will set this conf to hive.

SparkQA commented:
**[Test build #56382 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56382/consoleFull)** for PR 12485 at commit [`6e3c366`](https://github.com/apache/spark/commit/6e3c366a71f2dff1bbff59fd662b334fcc4798b8).

SparkQA commented:
**[Test build #56374 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56374/consoleFull)** for PR 12485 at commit [`9422128`](https://github.com/apache/spark/commit/9422128c6bd9c3ffbdf728eb6fd82eb17c50aef6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56374/
Test FAILed.

SparkQA commented:
**[Test build #56382 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56382/consoleFull)** for PR 12485 at commit [`6e3c366`](https://github.com/apache/spark/commit/6e3c366a71f2dff1bbff59fd662b334fcc4798b8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56382/
Test FAILed.

andrewor14 commented:
retest this please
